rero_ils/modules/stats/api/report.py

```diff
@@ -0,0 +1,445 @@
+                            doc_count = dist1[distrib2].buckets[key2].doc_count
+                        else:
+                            key2 = get_key2(dist2)
+                            doc_count = dist2.doc_count
+                        values[key2] = doc_count
+                        y_keys.add(key2)
+                if values:
+                    results[key1]['values'] = values
+            y_keys = sorted(y_keys)
+            x_keys = sorted(results.keys())
+        return self._process_distributions(x_keys, y_keys, results)
+
+    def _process_distributions(self, x_keys, y_keys, results):
+        """Process the elasticsearch aggregations results."""
+        data = []
+        if y_keys:
+            data.append([''] + y_keys)
+
+        for x_key in x_keys:
+            values = [x_key]
+            if y_keys:
+                for y_key in y_keys:
+                    values.append(
+                        results[x_key].get('values', {}).get(y_key, 0))
+            else:
+                values.append(results[x_key].get('count', 0))
+            data.append(values)
+        return data
+
+    def compute(self, force=False):
+        """Collect data for report.
+
+        :param force: compute even if the configuration is not active.
+        :returns: results data of report
+        """
+        if not self.is_active and not force:
+            return
+        cfg = self.es_config.get(self.indicator)
+        # base query
+        search = cfg['query']()
+        # no distributions returns the count
+        if not self.distributions:
+            return [[search.count()]]
+
+        # distributions
+        aggs = search.aggs
+
+        # compute distributions using aggregations
+        for dist in self.distributions:
+            aggs = aggs.bucket(dist, cfg['aggs'][dist]())
+
+        # compute the aggregations using es
+        results = search.execute()
+
+        return self._process_aggregations(results)
+
+    def _get_range_period(self, period):
+        """Get the range period for elasticsearch date range aggs."""
+        if period not in ['month', 'year']:
+            return None
+        if period == 'month':
+            # now - 1 month
+            previous_month = datetime.now() - relativedelta(months=1)
+            # get last day of previous month using relativedelta with
+            # day=31: this will add a max value of 31 days but stays
+            # in the same month
+            previous_month = previous_month + relativedelta(day=31)
+            month = '%02d' % previous_month.month
+            _from = f'{previous_month.year}-{month}-01T00:00:00'
+            _to = f'{previous_month.year}-{month}-{previous_month.day}'
+            _to = f'{_to}T23:59:59'
+        else:
+            previous_year = datetime.now().year - 1
+            _from = f'{previous_year}-01-01T00:00:00'
+            _to = f'{previous_year}-12-31T23:59:59'
+        return dict(gte=_from, lte=_to)
```

Comment:
Construction of this method seems strange for me. I would prefer simplier (for me and this is cosmetic):
```python
def _get_range_period(self, period):
        """Get the range period for elasticsearch date range aggs."""
        if period == 'month':
            # now - 1 month
            previous_month = datetime.now() - relativedelta(months=1)
            # get last day of previous month using relativedelta with
            # day=31: this will add a max value of 31 days but stays
            # in the same month
            previous_month = previous_month + relativedelta(day=31)
            year = '%04d' % previous_month.year
            month = '%02d' % previous_month.month
            day = '%02d' % previous_month.day
            _from = f'{year}-{month}-01T00:00:00'
            _to = f'{year}-{month}-{day}T23:59:59'
            return dict(gte=_from, lte=_to)
        elif period == 'year':
            previous_year = datetime.now().year - 1
            _from = f'{previous_year}-01-01T00:00:00'
            _to = f'{previous_year}-12-31T23:59:59'
            return dict(gte=_from, lte=_to)
```